openslides_backend/action/actions/user/user_mixin.py

```diff
@@ -173,3 +173,55 @@ def get_history_information(self) -> Optional[HistoryInformation]:
                     fqid_from_collection_and_id("user", instance["id"])
                 ] = instance_information
         return information
+
+
+class DuplicateCheckMixin(Action):
+    def init_duplicate_set(
+        self, usernames: List[str], names_and_emails: List[Any]
+    ) -> None:
+        # we don't want to call "get_all", so need to construct this large
+        # filter here.
+        filter_ = Or(
+            *[FilterOperator("username", "=", un) for un in usernames],
+            *[
+                And(
+                    FilterOperator("first_name", "=", entry[0]),
+                    FilterOperator("last_name", "=", entry[1]),
+                    FilterOperator("email", "=", entry[2]),
+                )
+                for entry in names_and_emails
+            ],
+        )
+
+        users = self.datastore.filter(
+            "user",
+            filter_,
+            ["id", "username", "first_name", "last_name", "email"],
+            lock_result=False,
+        ).values()
+
+        # for getting the ids in update case (username, names_and_email)
+        self.username_to_id = {values["username"]: values["id"] for values in users}
+        self.names_and_email_to_id = {
+            (
+                values.get("first_name"),
+                values.get("last_name"),
+                values.get("email"),
+            ): values["id"]
+            for values in users
+        }
+
+    def check_username_for_duplicate(self, username: str) -> bool:
+        result = username in self.username_to_id
+        if not result:
+            self.username_to_id[username] = None
+        return result
+
+    def check_name_and_email_for_duplicate(
+        self, first_name: Optional[str], last_name: Optional[str], email: Optional[str]
```

Comment:
If `username` is not provided, all other three fields are required (see https://github.com/OpenSlides/OpenSlides/wiki/without_templ.account.json_upload#user-matching), so these fields are not optional. This should also be checked somewhere in the code above.

Reply:
Done. I have added a check in json_upload and removed the Optional.